RZEffects/Core/RZEffectContext.m

```diff
@@ -0,0 +1,344 @@
+//
+//  RZEffectContext.m
+//
+//  Created by Rob Visentin on 2/18/15.
+//  Copyright (c) 2015 Raizlabs. All rights reserved.
+//
+
+#import <OpenGLES/ES2/glext.h>
+#import <OpenGLES/EAGL.h>
+#import <objc/runtime.h>
+
+#import "RZEffectContext.h"
+
+static GLuint RZCompileShader(const GLchar *source, GLenum type)
+{
+    GLuint shader = glCreateShader(type);
+    GLint length = (GLuint)strlen(source);
+
+    glShaderSource(shader, 1, &source, &length);
+    glCompileShader(shader);
+
+#if DEBUG
+    GLint success;
+    glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
+
+    if ( success != GL_TRUE ) {
+        GLint length;
+        glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &length);
+
+        GLchar *logText = malloc(length + 1);
+        logText[length] = '\0';
+        glGetShaderInfoLog(shader, length, NULL, logText);
+
+        fprintf(stderr, "Error compiling shader: %s\n", logText);
+
+        free(logText);
+    }
+#endif
+
+    return shader;
+}
+
+@interface RZEffectContext ()
+
+@property (strong, nonatomic, readwrite) dispatch_queue_t contextQueue;
+@property (strong, nonatomic) EAGLContext *glContext;
+
+@property (strong, nonatomic) NSMutableDictionary *compiledShaders;
+
+@property (assign, nonatomic) CVOpenGLESTextureCacheRef textureCache;
+
+@end
+
+@implementation RZEffectContext
+
+#pragma mark - lifecycle
+
++ (instancetype)defaultContext
+{
+    static id s_DefaultContext = nil;
+
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        s_DefaultContext = [[self alloc] init];
+    });
+
+    return s_DefaultContext;
+}
+
++ (RZEffectContext *)currentContext
+{
+    return objc_getAssociatedObject([EAGLContext currentContext], _cmd);
+}
+
+- (instancetype)init
+{
+    return [self initWithSharedContext:nil];
+}
+
+- (instancetype)initWithSharedContext:(RZEffectContext *)shareContext
+{
+    self = [super init];
+    if ( self ) {
+        const char *queueLabel = [NSString stringWithFormat:@"com.rzeffects.context-%lu", (unsigned long)self.hash].UTF8String;
+        _contextQueue = dispatch_queue_create(queueLabel, DISPATCH_QUEUE_SERIAL);
+
+        _glContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2 sharegroup:shareContext.glContext.sharegroup];
+        objc_setAssociatedObject(_glContext, @selector(currentContext), self, OBJC_ASSOCIATION_ASSIGN);
```

Comment:
Previously it was using an Apple-recommended way to use either an ES3 or ES2 context. But that way was bugged and never worked. So to be safe I converted to always use ES2. However.. the blur effect doesn't work right in ES2 contexts (separate bug). I want to re-introduce smartly picking the API, but in a way that works. I'll log a story